Fixes 92541 and 92562

craigtaverner opened the pull request:
An optimization in the GeoHexGridTiler has exceptions for polygons
with two bounding points within the polar cells, but not all points
within those cells. So when the cells are polar we do an additional
check for the other two bounds.

Fixes https://github.com/elastic/elasticsearch/issues/92541
Fixes https://github.com/elastic/elasticsearch/issues/92562

craigtaverner committed:
Fixes 92541 and 92562

An optimization in the GeoHexGridTiler has exceptions for polygons
with two bounding points within the polar cells, but not all points
within those cells. The normal hexRing neighbour search is unsufficient
due to planar projection distortions.

We found that just one additional next-neighbour cell was needed for
each polar cell to maintain the optimization.

Fixes https://github.com/elastic/elasticsearch/issues/92541
Fixes https://github.com/elastic/elasticsearch/issues/92562

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner commented:
Changed label to non-issue (from bug) because this is a fix to an unreleased feature (geohex over geoshape). And force pushed a rebase back to before the changelog yaml.

iverase commented:
I agree that the issue is with polar cells but not sure if the fix is enough, only for level 0 polar cells. I had a quick look and I was able to reproduce the same error for a polar cells at level 1:

```
public void testPolarCellLevel1() throws Exception {
        int precision = 1;
        String polygon = "BBOX (-84.24596376729815, 43.36113427778119, 90.0, 83.51476833522361)";
        Geometry geometry = WellKnownText.fromWKT(StandardValidator.instance(true), true, polygon);
        GeoShapeValues.GeoShapeValue value = geoShapeValue(geometry);
        GeoShapeCellValues unboundedCellValues = new GeoShapeCellValues(
            makeGeoShapeValues(value),
            getUnboundedGridTiler(precision),
            NOOP_BREAKER
        );

        assertTrue(unboundedCellValues.advanceExact(0));
        int numBuckets = unboundedCellValues.docValueCount();
        int expected = expectedBuckets(value, precision, null);
        assertThat("[" + precision + "] bucket count", numBuckets, equalTo(expected));
    }
```

I am thinking it might be simpler to disable this optimisations whenever the cells are polar?

craigtaverner committed:
Fixes 92541 and 92562

An optimization in the GeoHexGridTiler has exceptions for polygons
with two bounding points within the polar cells, but not all points
within those cells. So when the cells are polar we do an additional
check for the other two bounds.

Fixes https://github.com/elastic/elasticsearch/issues/92541
Fixes https://github.com/elastic/elasticsearch/issues/92562

craigtaverner committed:
Remove unused imports

craigtaverner committed:
Missing highest resolution for poles

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java:
```diff
@@ -19,6 +19,15 @@
 
     private static final long[] RES0CELLS = H3.getLongRes0Cells();
+    // The hexRing neighbours optimization is insufficient at the Poles, so we do a little extra checking
+    private static final long[] NORTH_POLAR_CELLS = new long[H3.MAX_H3_RES];
```
I think this information is already in H3CartesianUtils, cannot we add just methods on that class?

iverase replied:
Actually there is a method -> H3CartesianUtils.#isPolar that does what you want.

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java:
```diff
@@ -48,10 +57,9 @@ public int setValues(GeoShapeCellValues values, GeoShapeValues.GeoShapeValue geo
-            final long maxH3 = H3.geoToH3(bounds.maxY(), bounds.maxX(), precision);
-            if (minH3 == maxH3) {
-                return setValuesFromPointResolution(minH3, values, geoValue);
+            final long singleCell = boundsInSameCell(bounds, precision);
```
Can we add here the constraint that precision > 0. We are doing that check later on.

iverase reviewed (approved):
LGTM. 

I think we should use the information of the NORTH / SOUTH cells from H#CartesianUtils, but otherwise I like the approach.

craigtaverner committed:
Use H3CartesianUtil.isPolar()
